### PR TITLE
Improve reduction core

### PR DESCRIPTION
**Pull request by okuta**
This PR adds `_get_reduction_args` to remove duplicated code.
~Please merge #2689 before this PR.~

### CODE REVIEW

**Comment by asi1024** [comment]
Sorry for my late review. Could you resolve conflicts?

**Comment by okuta** [comment]
I rebased.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by asi1024** [comment]
This PR is just for code simplification? I compared the performance between this branch and by using [cupy-perf](https://github.com/niboshi/cupy-perf), but I saw no performance improvement.

**Comment by okuta** [comment]
This PR focuses simplify.
I want to send performance improvement PR after this PR.

File: cupy/core/reduction.pxi [hunk 1]
```diff
@@ -114,63 +114,93 @@ cpdef tuple _get_out_shape(
...
+                raise NotImplementedError('Illegal conditions')
+        for i, a in enumerate(args):
+            if isinstance(a, ndarray):
+                args[i] = _manipulation._transpose(a, axis_permutes)
         shape = tuple([shape[i] for i in axis_permutes])
+    return shape
 
+
+cdef Py_ssize_t _get_contiguous_size(
+        list args, tuple params, Py_ssize_t ndim, Py_ssize_t out_ndim) except*:
```
Comment (asi1024) [inline cupy/core/reduction.pxi]: This function returns non-negative value, so `except -1` seems appropriate.
```suggestion
        list args, tuple params, Py_ssize_t ndim, Py_ssize_t out_ndim) except -1:
```

File: cupy/core/reduction.pxi [hunk 2]
```diff
@@ -114,63 +114,93 @@ cpdef tuple _get_out_shape(
     return tuple([shape[i] for i in out_axis])
 
 
-cpdef tuple _get_permuted_args(
-        list args, tuple axis_permutes, tuple shape, tuple params,
-        Py_ssize_t out_ndim):
+cdef tuple _set_permuted_args(
+        list args, tuple axis_permutes, tuple shape, tuple params):
```
Comment (asi1024) [inline cupy/core/reduction.pxi]: I feel a little confused that this function updates `args` internally. Could you add a comment?

File: cupy/core/reduction.pxi (hunk 1 above)
Comment (asi1024) [reply cupy/core/reduction.pxi]: BTW, I think that this function never raises any exception.

File: cupy/core/reduction.pxi (hunk 2 above)
Comment (okuta) [reply cupy/core/reduction.pxi]: I added comment.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**